### ISSUE TITLE
Gapless

### ISSUE THREAD

**Issue by dnpawate** [google/ExoPlayer#4425]
Hi

Need a simple confirmation
1. For lossy format say MP3 will android framework will take care of removing gaps between files 
2. Do we need to design android player which will always skip last one second of MP3 songs
3. Do we need hardware support to achieve gapless playback
4. If we have long silence of 8 sec at end of file, who's responsibility to detect the silence android framework 
5. Does only lame tag in mp3 file will only supports gapless playback

Please clarify your justification in detail.

Thanks a lot

**Comment by andrewlewis** [comment]
1. ExoPlayer supports playing MP3s gaplessly if they have the required encoder delay/padding metadata stored in iTunes-style ID3 metadata or in a Xing header. In recent versions of ExoPlayer the encoder delay/padding are stripped by ExoPlayer in `TrimmingAudioProcessor`, so we don't rely on the media framework for this. (The media framework does support removing the silent samples via format keys set on `MediaCodec` but we don't take advantage of that as we need to support non-MediaCodec-based audio renderers too.)
2. I'm not sure I understand the question. Please could you rephrase it?
3. No special hardware support is needed.
4. The number of encoder delay/padding samples to strip is normally signaled via extra information stored in the container, so it's up to the content producer to include that. I think standard tools like lame tend to include the required metadata. ExoPlayer doesn't support automatically working out how many samples to remove when concatenating audio streams (it does support skipping silence but this is intended for speeding up spoken content, and won't give perfectly seamless transitions between music tracks).
5. See answer (1).

**Comment by dnpawate** [comment]
Hi andrew,

Thank you for reply

1. each gapless applications in playstore will they use exoplayer or android media framework 
2. even though gapless is enabled i see with most of the mp3 files do not skip silence
just not able to figure out, why there are no seamless gapless operations

**Comment by andrewlewis** [comment]
1. They could be using either. I think ExoPlayer and the Android media framework support the same types of gapless metadata. Apps can also do something completely custom, of course.
2. You need to make sure the MP3s are authored with the correct metadata (an ID3 metadata comment or Xing header as mentioned above). If the metadata is present then gapless playback with `ConcatenatingMediaSource` should just work.
